ic-asset/src/asset_config.rs

```
@@ -0,0 +1,627 @@
+        ))
+    }
+}
+
+impl AssetConfigRule {
+    fn applies(&self, cannonical_path: &Path) -> bool {
+        self.r#match.is_match(cannonical_path)
+    }
+}
+
+impl AssetConfig {
+    fn merge(mut self, other: Self) -> Self {
+        if let Some(c) = other.cache {
+            self.cache = Some(c);
+        };
+        match (self.headers.as_mut(), other.headers) {
+            (Some(sh), Some(oh)) => sh.extend(oh),
+            (_, oh) => self.headers = oh,
+        };
+        self
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct AssetSourceDirectoryConfiguration {
+    config_map: Map,
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize)]
+pub(crate) struct AssetConfig {
+    pub(crate) cache: Option<CacheConfig>,
+    pub(crate) headers: Option<HeadersConfig>,
+}
+
+impl Default for AssetConfig {
+    fn default() -> Self {
+        Self {
+            headers: Some(HashMap::new()),
+            cache: Some(CacheConfig::default()),
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+struct AssetConfigTreeNode {
+    pub parent: Option<Arc<AssetConfigTreeNode>>,
+    pub rules: Vec<AssetConfigRule>,
+}
+
+impl AssetSourceDirectoryConfiguration {
+    pub(crate) fn load(root_dir: &Path) -> anyhow::Result<Self> {
+        let mut config_map = HashMap::new();
+        AssetConfigTreeNode::load(None, &root_dir, &mut config_map)?;
+        Ok(Self { config_map })
+    }
+
+    pub(crate) fn get_asset_config(&self, canonical_path: &Path) -> AssetConfig {
+        self.config_map
+            .get(canonical_path.parent().unwrap())
+            .unwrap()
+            .get_config(canonical_path)
+    }
+}
+
+impl AssetConfigTreeNode {
+    fn load(
+        parent: Option<Arc<AssetConfigTreeNode>>,
+        dir: &Path,
+        configs: &mut HashMap<PathBuf, Arc<AssetConfigTreeNode>>,
+    ) -> anyhow::Result<()> {
+        let mut rules = vec![];
+        if let Ok(file) = File::open(dir.join(ASSETS_CONFIG_FILENAME)) {
```

If an .ic-assets.json file exists, but is missing read permissions, would you expect `load` to report an error, or silently skip the file?

---

I'd expect it to fail, but that's not how it works atm. Kinda struggling to come up with tests for that - manually setting config file permission to `000` doesn't seem to do the trick...